src/main/java/io/cryostat/net/security/jwt/AssetJwtHelper.java

```diff
@@ -0,0 +1,155 @@
+
+import java.net.MalformedURLException;
+import java.net.SocketException;
+import java.net.URISyntaxException;
+import java.net.UnknownHostException;
+import java.text.ParseException;
+import java.time.Instant;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.Set;
+
+import io.cryostat.core.log.Logger;
+import io.cryostat.net.web.WebServer;
+
+import com.nimbusds.jose.EncryptionMethod;
+import com.nimbusds.jose.JOSEException;
+import com.nimbusds.jose.JWEAlgorithm;
+import com.nimbusds.jose.JWEDecrypter;
+import com.nimbusds.jose.JWEEncrypter;
+import com.nimbusds.jose.JWEHeader;
+import com.nimbusds.jose.JWEObject;
+import com.nimbusds.jose.JWSAlgorithm;
+import com.nimbusds.jose.JWSHeader;
+import com.nimbusds.jose.JWSSigner;
+import com.nimbusds.jose.JWSVerifier;
+import com.nimbusds.jose.Payload;
+import com.nimbusds.jwt.JWT;
+import com.nimbusds.jwt.JWTClaimsSet;
+import com.nimbusds.jwt.SignedJWT;
+import com.nimbusds.jwt.proc.BadJWTException;
+import com.nimbusds.jwt.proc.DefaultJWTClaimsVerifier;
+import dagger.Lazy;
+
+public class AssetJwtHelper {
+
+    public static final String RESOURCE_CLAIM = "resource";
+    public static final String JMXAUTH_CLAIM = "jmxauth";
+
+    private final Lazy<WebServer> webServer;
+    private final JWSSigner signer;
+    private final JWSVerifier verifier;
+    private final JWEEncrypter encrypter;
+    private final JWEDecrypter decrypter;
+    private final boolean subjectRequired;
+
+    AssetJwtHelper(
+            Lazy<WebServer> webServer,
+            JWSSigner signer,
+            JWSVerifier verifier,
+            JWEEncrypter encrypter,
+            JWEDecrypter decrypter,
+            boolean subjectRequired,
+            Logger logger) {
+        this.webServer = webServer;
+        this.signer = signer;
+        this.verifier = verifier;
+        this.encrypter = encrypter;
+        this.decrypter = decrypter;
+        this.subjectRequired = subjectRequired;
+    }
+
+    public String createAssetDownloadJwt(String subject, String resource, String jmxauth)
+            throws JOSEException, SocketException, UnknownHostException, URISyntaxException,
+                    MalformedURLException {
+        String issuer = webServer.get().getHostUrl().toString();
+        Date now = Date.from(Instant.now());
+        Date expiry = Date.from(now.toInstant().plusSeconds(120));
+        JWTClaimsSet claims =
+                new JWTClaimsSet.Builder()
+                        .issuer(issuer)
+                        .audience(issuer)
+                        .issueTime(now)
+                        .notBeforeTime(now)
+                        .expirationTime(expiry)
+                        .subject(subject)
+                        .claim(RESOURCE_CLAIM, resource)
+                        .claim(JMXAUTH_CLAIM, jmxauth)
+                        .build();
+
+        SignedJWT jwt = new SignedJWT(new JWSHeader.Builder(JWSAlgorithm.HS256).build(), claims);
```

Comment:
I was working on setting it up to re-use the existing SSL/TLS certs for asymmetric encryption but ran into several hurdles, and eventually got to the point that I questioned why it needed to be asymmetric - the token is produced by Cryostat and only needs to be readable by Cryostat, and not as it's passed to and from the client in between. It might be a nice-to-have feature for troubleshooting if the token was asymmetrically encrypted so that an end user could decrypt the tokens on their own and read what contents were encoded into it, but that didn't seem like a particularly pressing need. And, I really wanted to have the tokens signed and encrypted in all cases so that the tokens can't leak JMX credentials that are encoded into them, even if Cryostat did not have its own SSL/TLS certs for asymmetric encryption and was relying on an SSL proxy (ex. OpenShift Route edge termination), so that was another reason to just hardcode it to use a symmetric algorithm and a secret that it always generates at runtime regardless of the HTTPS setup or other factors.

Reply:
Makes sense, thanks